Support huggingface reading and writing for multi rank case

ankitageorge opened the pull request:
Summary: This diff adds the ability for HF reader/writer to read/write in a distributed way. We do this by sending all the tensors meant for the same file to the same rank.

Test Plan:
ensure existing tests pass
I also ran a full end to end test on my devserver to read/write from my HF repo

Differential Revision: D70096439




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0

fegin commented on torch/distributed/checkpoint/_hf_planner.py:
```diff
@@ -0,0 +1,53 @@
+
+class _HuggingFaceLoadPlanner(DefaultLoadPlanner):
+    """
+    A save planner that dedups the save plans based on the fqn to file index mapping.
```
a load planner?

joecummings commented on torch/distributed/checkpoint/_hf_storage.py:
```diff
@@ -172,8 +179,14 @@ def read_data(self, plan: LoadPlan, planner: LoadPlanner) -> Future[None]:
-                    target_tensor = planner.resolve_tensor(req).detach()
-                    target_tensor.resize_(tensor.size())
+                    target_tensor = planner.resolve_tensor(req)
+                    if isinstance(planner, _HuggingFaceLoadPlanner) and planner.allow_tensor_resize:
```
Is allowing tensor resizing a generic enough request that it could be part of the DefaultLoadPLanner and then you wouldn't need to check if the planner was an instance of the `_HuggingFaceLoadPlanner`?

joecummings reviewed (approved):
This looks good to me, just a general comment

ankitageorge committed:
Support huggingface reading and writing for multi rank case (#148189)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/148189

This diff adds the ability for HF reader/writer to read/write in a distributed way. We do this by sending all the tensors meant for the same file to the same rank.

Test Plan:
ensure existing tests pass
I also ran a full end to end test on my devserver to read/write from my HF repo

Reviewed By: saumishr

Differential Revision: D70096439
